File: python/socketd/socketd/transport/client/ClientChannel.py
import asyncio
import traceback
from asyncio import Future

from socketd.exception.SocketDExecption import SocketDException, SocketDChannelException
from socketd.transport.client.Client import ClientInternal
from socketd.transport.core.Asserts import Asserts
from socketd.transport.core.ChannelInternal import ChannelInternal
from socketd.transport.core.Costants import Constants
from socketd.utils.LogConfig import log
from socketd.transport.core.impl.SessionDefault import SessionDefault
from socketd.transport.stream.Stream import StreamInternal
from socketd.transport.core.impl.ChannelBase import ChannelBase
from socketd.transport.client.ClientConnector import ClientConnector

from socketd.transport.client.ClientHeartbeatHandler import ClientHeartbeatHandlerDefault
from socketd.utils.sync_api.AtomicRefer import AtomicRefer


class ClientChannel(ChannelBase):
    def __init__(self, client: ClientInternal, connector: ClientConnector):
        super().__init__(connector.get_config())
        self.__client = client
        self.__connector: ClientConnector = connector
        self.__sessionShell = SessionDefault(self)

        self.__real: ChannelInternal | None = None
        self.__heartbeatScheduledFuture: Future | None = None

        if client.get_heartbeat_handler():
            self.__heartbeatHandler: type[ClientHeartbeatHandlerDefault] = client.get_heartbeat_handler()
        else:
            self.__heartbeatHandler = ClientHeartbeatHandlerDefault

        self.__isConnecting = AtomicRefer(False)

        self.init_heartbeat()

    def __del__(self):
        try:
            if not self.__heartbeatScheduledFuture.done():
                self.__heartbeatScheduledFuture.cancel()
        except Exception as e:
            e_msg = traceback.format_exc()
            log.warning(e_msg)

    async def __heartbeatScheduled(self) -> None:
        while True:
            await asyncio.sleep(self.__client.get_heartbeat_interval() / 1000)
            await self.heartbeat_handle()

    def init_heartbeat(self):
        if self.__heartbeatScheduledFuture:
            self.__heartbeatScheduledFuture.cancel()

        if self.__connector.auto_reconnect():
            self.__heartbeatScheduledFuture = asyncio.create_task(self.__heartbeatScheduled())

    async def heartbeat_handle(self):
        if self.__real:
            if self.__real.get_handshake() is None:
                return

            if Asserts.is_closed_and_end(self.__real):
                log.debug("Client channel is closed (pause heartbeat), sessionId=" + self.get_session().session_id())
                await self.close(self.__real.close_code())
                return

            if self.__real.is_closing():
                return

        with self:
            try:
                await self.internalCheck()
                await self.__heartbeatHandler(self.get_session())
            except Exception as e:
                if self.__connector.auto_reconnect():
                    if self.__real:
                        await self.__real.close(code=Constants.CLOSE2001_ERROR)
                    self.__real = None
                raise e

    def is_valid(self):
        if self.__real is None:
            return False
        else:
            return self.__real.is_valid()

    def close_code(self):
        if self.__real is None:
            return False
        else:
            return self.__real.close_code()

    def get_remote_address(self):
        if self.__real is None:
            return None
        else:
            return self.__real.get_remote_address()

    def get_local_address(self):
        if self.__real is None:
            return None
        else:
            return self.__real.get_local_address()

    async def send(self, frame, stream: StreamInternal):
        Asserts.assert_closed_and_end(self.__real)

        try:
            await self.internalCheck()

            if self.__real is None:
                # 销毁心跳任务
                self.__heartbeatScheduledFuture.cancel()
                raise SocketDChannelException("Client channel is not connected")

            await self.__real.send(frame, stream)
        except SocketDException as s:
            raise s
        except Exception as e:
            if self.__connector.auto_reconnect():
                await self.internalCloseIfError()

            raise SocketDChannelException(f"Client channel send failed {e}")

    async def retrieve(self, frame, stream: StreamInternal):
        await self.__real.retrieve(frame, stream)

    async def close(self, code):
        try:
            if self.__heartbeatScheduledFuture:
                self.__heartbeatScheduledFuture.cancel()

            if self.__connector:
                await self.__connector.close()

            if self.__real:
                await self.__real.close(code)
        except Exception as e:
            log.error(e)
        finally:
            await super().close(code)

    def get_session(self):
        return self.__sessionShell

    def on_error(self, error: Exception):
        self.__real.on_error(error)

    async def reconnect(self):
        self.init_heartbeat()
        await self.internalCheck()

    @log.catch
    async def connect(self):
        with self.__isConnecting as isConnected:
            if isConnected:
                return
            else:
                self.__isConnecting.set(True)

        try:
            if self.__real:
                await self.__real.close(Constants.CLOSE2002_RECONNECT)

            self.__real = await self.__client.get_connect_handler()(self.__connector)
            self.__real.set_session(self.__sessionShell)
            self.set_handshake(self.__real.get_handshake())
        except TimeoutError as t:
            log.error(f"socketD connect timed out: {t}")
        except Exception as e:
            log.error(e)
            raise SocketDChannelException(f"socketD connect")
        finally:
            self.__isConnecting.set(False)

    async def internalCloseIfError(self):
        if self.__real:
            await self.__real.close(Constants.CLOSE2001_ERROR)
            self.__real = None

    async def internalCheck(self):
        if self.__real is None or self.__real.is_valid() == False:
            await self.connect()
            return True
        else:
            return False

    def is_closing(self) -> bool:
        return self.__real.is_closing() if self.__real else 0

    def get_live_time(self) -> int:
        return self.__real.get_live_time() if self.__real else 0

File: javascript/socketd/src/socketd/transport/client/ClientChannel.ts
import {Channel, ChannelInternal} from "../core/Channel";
import type { Frame } from "../core/Frame";
import type { Session } from "../core/Session";
import type { StreamInternal } from "../stream/Stream";
import type { ClientConnector } from "./ClientConnector";
import { Constants } from "../core/Constants";
import { Asserts } from "../core/Asserts";
import { SocketDChannelException, SocketDException } from "../../exception/SocketDException";
import {RunUtils} from "../../utils/RunUtils";
import {SessionDefault} from "../core/SessionDefault";
import {SocketAddress} from "../core/SocketAddress";
import {ClientHeartbeatHandler, ClientHeartbeatHandlerDefault} from "./ClientHeartbeatHandler";
import {ClientInternal} from "./Client";
import {ClientConnectHandler, ClientConnectHandlerDefault} from "./ClientConnectHandler";
import {ChannelBase} from "../core/impl/ChannelBase";

/**
 * 客户端通道
 *
 * @author noear
 * @since 2.0
 */
export class ClientChannel extends ChannelBase implements Channel {
    //客户端
    private _client: ClientInternal;
    //连接器
    private _connector: ClientConnector;
    //会话壳
    private _sessionShell: Session;
    //真实通道
    private _real: ChannelInternal | null;
    //连接处理
    private _connectHandler : ClientConnectHandler;
    //心跳处理
    private _heartbeatHandler: ClientHeartbeatHandler;
    //心跳调度
    private _heartbeatScheduledFuture: any;
    //连接状态
    private _isConnecting: boolean = false;

    constructor(client: ClientInternal, connector: ClientConnector) {
        super(connector.getConfig());
        this._client = client;
        this._connector = connector;
        this._sessionShell = new SessionDefault(this);

        this._connectHandler = new ClientConnectHandlerDefault(client.getConnectHandler());
        this._heartbeatHandler = new ClientHeartbeatHandlerDefault(client.getHeartbeatHandler());

        this.initHeartbeat();
    }

    /**
     * 初始化心跳（关闭后，手动重链时也会用到）
     */
    initHeartbeat() {
        if (this._heartbeatScheduledFuture) {
            clearInterval(this._heartbeatScheduledFuture);
        }

        if (this._connector.autoReconnect()) {
            this._heartbeatScheduledFuture = setInterval(async () => {
                try {
                    await this.heartbeatHandle();
                } catch (e) {
                    console.debug("Client channel heartbeat failed: {link=" + this._connector.getConfig().getLinkUrl() + "}");
                }
            }, this._client.getHeartbeatInterval());
        }
    }

    /**
     * 心跳处理
     */
    async heartbeatHandle() {
        if (this._real) {
            //说明握手未成
            if (this._real.getHandshake() == null) {
                return;
            }

            //关闭并结束了
            if (Asserts.isClosedAndEnd(this._real)) {
                console.debug(`Client channel is closed (pause heartbeat), sessionId=${this.getSession().sessionId()}`);
                //可能是被内层的会话关闭的，跳过了外层
                this.close(this._real.closeCode());
                return;
            }

            //或者正在关闭中
            if (this._real.isClosing()) {
                return;
            }
        }

        try {
            await this.internalCheck();

            this._heartbeatHandler.clientHeartbeat(this.getSession());
        } catch (e) {
            if (e instanceof SocketDException) {
                throw e;
            }

            if (this._connector.autoReconnect()) {
                this.internalCloseIfError();
            }

            throw e;
        }
    }


    /**
     * 是否有效
     */
    isValid() {
        if (this._real == null) {
            return false;
        } else {
            return this._real.isValid();
        }
    }

    /**
     * 是否关闭中
     */
    isClosing() {
        if (this._real == null) {
            return false;
        } else {
            return this._real.isClosing();
        }
    }

    /**
     * 是否已关闭
     */
    closeCode(): number {
        if (this._real == null) {
            return 0;
        } else {
            return this._real.closeCode();
        }
    }

    getLiveTime(): number {
        if (this._real) {
            return this._real!.getLiveTime();
        } else {
            return 0;
        }
    }

    getRemoteAddress(): SocketAddress | null {
        if (this._real) {
            return  this._real.getRemoteAddress();
        } else {
            return null;
        }
    }

    getLocalAddress(): SocketAddress | null {
        if (this._real) {
            return  this._real.getLocalAddress();
        } else {
            return null;
        }
    }

    /**
     * 发送
     *
     * @param frame  帧
     * @param stream 流（没有则为 null）
     */
    send(frame: Frame, stream: StreamInternal<any>) {
        Asserts.assertClosedAndEnd(this._real);

        this.internalCheck().then(res => {
            if (this._real) {
                try {
                    this._real!.send(frame, stream);
                } catch (err) {
                    if (stream) {
                        // @ts-ignore
                        stream.onError(err);
                    }
                }
            } else {
                //有可能此时仍未连接
                const err = new SocketDChannelException("Client channel is not connected");
                if (stream) {
                    stream.onError(err);
                }
            }
        }, err => {
            if (this._connector.autoReconnect()) {
                this.internalCloseIfError();
            }

            if (stream) {
                stream.onError(err);
            }
        });
    }

    async reconnect() {
        this.initHeartbeat();

        await this.internalCheck();
    }

    onError(error: any) {
        this._real!.onError(error);
    }

    close(code: number) {
        RunUtils.runAndTry(() => clearInterval(this._heartbeatScheduledFuture));
        RunUtils.runAndTry(() => this._connector.close());
        if (this._real) {
            RunUtils.runAndTry(() => this._real!.close(code));
        }
        super.close(code);
    }

    getSession(): Session {
        return this._sessionShell;
    }

    async connect() {
        if (this._isConnecting) {
            return;
        } else {
            this._isConnecting = true;
        }

        try {
            if (this._real != null) {
                this._real.close(Constants.CLOSE2002_RECONNECT);
            }

            this._real = await this._connectHandler.clientConnect(this._connector);
            //原始 session 切换为带壳的 session
            this._real.setSession(this._sessionShell);
            //同步握手信息
            this.setHandshake(this._real.getHandshake());
        } finally {
            this._isConnecting = false;
        }
    }

    private internalCloseIfError() {
        if (this._real != null) {
            this._real.close(Constants.CLOSE2001_ERROR);
            this._real = null;
        }
    }

    /**
     * 预备检测
     *
     * @return 是否为新链接
     */
    private async internalCheck(): Promise<boolean> {
        if (this._real == null || this._real.isValid() == false) {
            await this.connect();

            return true;
        } else {
            return false;
        }
    }
}
